modules/light-clients/08-wasm/types/migrate_contract.go

```diff
@@ -0,0 +1,41 @@
+package types
+
+import (
+	"bytes"
+	"encoding/hex"
+
+	errorsmod "cosmossdk.io/errors"
+	storetypes "cosmossdk.io/store/types"
+
+	"github.com/cosmos/cosmos-sdk/codec"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+)
+
+// MigrateContract calls the migrate entry point on the contract with the given
+// migrateMsg. The contract must exist and the code hash must be found in the
+// store. If the code hash is the same as the current code hash, an error is returned.
+// This does not update the code hash in the client state.
+func (cs ClientState) MigrateContract(
+	ctx sdk.Context, cdc codec.BinaryCodec, clientStore storetypes.KVStore,
+	clientID string, newCodeHash, migrateMsg []byte,
+) error {
+	if !HasCodeHash(ctx, newCodeHash) {
+		return ErrWasmCodeHashNotFound
+	}
+
+	if bytes.Equal(cs.CodeHash, newCodeHash) {
+		return errorsmod.Wrapf(ErrWasmCodeExists, "new code hash (%s) is the same as current code hash (%s)", hex.EncodeToString(newCodeHash), hex.EncodeToString(cs.CodeHash))
+	}
+
+	// update the code hash, this needs to be done before the contract migration
+	// so that wasmMigrate can call the right code. Note that this is not
+	// persisted to the client store.
+	cs.CodeHash = newCodeHash
```

Comment:
Okay, so we call migrate on the new code hash and the new code hash is given access to the old code hashes client store? If the new contract needed the specific code hash of the old contract, it could query it via the client state?

Reply:
Yes, it is given the access to clientStore so that it can perform any state migration if needed! The new code should have no reason to need the old code hash. If it is needed for any reason, it is already in the clientState. Or it could be a built in constant in the new code since it is a future iteration of the same code.